spdmlib/src/crypto/x509v3.rs

```diff
@@ -0,0 +1,1628 @@
+            match ai.cmp(bi) {
+                core::cmp::Ordering::Equal => continue,
+                _ => return false,
+            }
+        }
+        true
+    }
+}
+
+#[cfg(all(test,))]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_case0_object_identifiers_are_same() {
+        let lt = [0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0bu8];
+        let lt_wrong1 = [0x2b, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0bu8];
+        let lt_wrong2 = [0x2b, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0b, 0xb0u8];
+        let lt_empty: [u8; 0] = [];
+        assert!(object_identifiers_are_same(&lt, OID_RSA_SHA256RSA));
+        assert!(!object_identifiers_are_same(&lt, OID_RSA_SHA384RSA));
+        assert!(!object_identifiers_are_same(&lt_wrong1, OID_RSA_SHA256RSA));
+        assert!(!object_identifiers_are_same(&lt_wrong2, OID_RSA_SHA256RSA));
+        assert!(!object_identifiers_are_same(&lt_empty, OID_RSA_SHA384RSA));
+    }
+
+    #[test]
+    fn test_case0_get_oid_by_base_asym_algo() {
+        assert_eq!(
+            get_oid_by_base_asym_algo(SpdmBaseAsymAlgo::TPM_ALG_RSASSA_2048),
+            Some(OID_RSA_SHA256RSA)
+        );
+        assert_eq!(
+            get_oid_by_base_asym_algo(SpdmBaseAsymAlgo::TPM_ALG_ECDSA_ECC_NIST_P256),
+            Some(OID_ECDSA_SHA256)
+        );
+    }
+
+    #[test]
+    fn test_case0_check_and_skip_common_tag() {
+        let sq1 = [
+            0x03, 0x68, 0x00, 0x30, 0x65, 0x02, 0x31, 0x00, 0xD7, 0x9C, 0x7F, 0x26, 0x91, 0x34,
+            0xA5, 0x2B, 0x79, 0xEA, 0x66, 0x15, 0x00, 0x88, 0x0A, 0x4D, 0xE7, 0xAD, 0x71, 0xC6,
+            0x2E, 0xE4, 0x7E, 0x37, 0xE1, 0x86, 0xEB, 0xE8, 0x55, 0xB0, 0x2F, 0xC5, 0xF3, 0xA9,
+            0xE0, 0x90, 0xF9, 0x0B, 0x82, 0xC5, 0xDF, 0x4A, 0x35, 0x9A, 0x0D, 0x35, 0x38, 0x4B,
+            0x02, 0x30, 0x40, 0xA7, 0xFE, 0x70, 0x39, 0x7B, 0x4B, 0xD7, 0xC2, 0x28, 0x72, 0x93,
+            0x93, 0x0C, 0x62, 0x12, 0x14, 0xF0, 0x70, 0x74, 0x0F, 0xFC, 0xB1, 0x21, 0x60, 0x40,
+            0x6D, 0x13, 0xA3, 0x59, 0x0E, 0x27, 0x06, 0xC1, 0x73, 0x4E, 0xCA, 0x40, 0x4C, 0x2D,
+            0xF5, 0x96, 0x48, 0x66, 0x05, 0xB1, 0xA6, 0x08,
+        ];
+        let sq2 = [0xA0, 0x03, 0x02, 0x01, 0x02];
+        let sq3 = [0x01, 0x01, 0xFF];
+        let sq4 = [0x01, 0x01, 0xFF, 0xAA];
+        let sq1_wrong = [0x01, 0x02, 0xFF];
+        assert_eq!(check_and_skip_common_tag(&sq1), Ok(106));
+        assert_eq!(check_and_skip_common_tag(&sq2), Ok(5));
+        assert_eq!(check_and_skip_common_tag(&sq3), Ok(3));
+        assert_eq!(check_and_skip_common_tag(&sq4), Ok(3));
+        assert_eq!(
+            check_and_skip_common_tag(&sq1_wrong),
+            Err(SPDM_STATUS_VERIF_FAIL)
+        );
+    }
+
+    #[test]
+    fn test_case0_check_and_skip_common_sequence() {
+        let sq1 = [
+            0x30, 0x10, 0x06, 0x07, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01, 0x06, 0x05, 0x2B,
+            0x81, 0x04, 0x00, 0x22,
+        ];
+        let sq2 = [
```

Comment:
please dont hardcode the certificate.

Reply:
> please dont hardcode the certificate.
> 
> See: #660, and #683.

Got it.
The hardcode is now removed, please check latest commit.